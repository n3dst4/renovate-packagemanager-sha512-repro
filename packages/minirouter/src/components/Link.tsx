import React, {
  ComponentProps,
  forwardRef,
  PropsWithChildren,
  useCallback,
} from "react";

import { AnyStep, DirectionType } from "../types";
import { useNavigationContext } from "../useNavigationContext";

type LinkProps = ComponentProps<"a"> &
  PropsWithChildren<{
    from?: DirectionType;
    to?: AnyStep | AnyStep[] | "up";
  }>;

// thanks to TS 5.5 this is a type guard even without the `: x is y` syntax
function isDirection(x: DirectionType) {
  return typeof x !== "string";
}

/**
 * Create a link
 *
 * @param from The direction to link from. This can be a "here", "root", or a
 * Direction.
 * @param to The step or steps to link to, or "up" to go up a level.
 */
export const Link = forwardRef<HTMLAnchorElement, LinkProps>(
  ({ children, from = "here", to = [], ...otherProps }, ref) => {
    const { navigate, parentSteps, currentStep } = useNavigationContext();
    if (isDirection(from)) {
      if (
        currentStep?.direction !== from &&
        !parentSteps.some((s) => s.direction === from)
      ) {
        throw new Error(
          `Link has "from" set to ${from.description} but the current step is not a descendant of that step`,
        );
      }
    }

    const followLink = useCallback(
      () => navigate(from, to),
      [from, navigate, to],
    );

    const onKeyDown = useCallback<
      React.KeyboardEventHandler<HTMLAnchorElement>
    >(
      (e) => {
        if (e.key === "Enter" || e.key === " ") {
          e.preventDefault();
          followLink();
        }
      },
      [followLink],
    );

    const onClick = useCallback<React.MouseEventHandler<HTMLAnchorElement>>(
      (e) => {
        e.preventDefault();
        followLink();
      },
      [followLink],
    );

    return (
      <a
        tabIndex={0}
        onClick={onClick}
        onKeyDown={onKeyDown}
        ref={ref}
        {...otherProps}
      >
        {children}
      </a>
    );
  },
);

Link.displayName = "Link";
