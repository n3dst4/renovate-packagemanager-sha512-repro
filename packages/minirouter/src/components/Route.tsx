import React from "react";

import { PropsWithChildrenAndDirection } from "../types";
import { useRoute } from "../useRoute";

/**
 *
 */
export const Route: React.FC<PropsWithChildrenAndDirection> = ({
  direction,
  children,
}) => {
  const result = useRoute({ direction, children });
  return result;
};
