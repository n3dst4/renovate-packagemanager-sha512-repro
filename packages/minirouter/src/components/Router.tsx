import React, { PropsWithChildren, useMemo, useState } from "react";

import { NavigationContext } from "../NavigationContext";
import { AnyStep, NavigationContextValue } from "../types";

function deepEquals(a: any, b: any) {
  return JSON.stringify(a) === JSON.stringify(b);
}

export const Router: React.FC<PropsWithChildren> = ({ children }) => {
  const [currentPath, setCurrentPath] = useState<AnyStep[]>([]);
  const navigationContextValue = useMemo<NavigationContextValue>(
    () => ({
      navigate: (from, to) => {
        const isUp = to === "up";
        const toArray = isUp ? [] : Array.isArray(to) ? to : [to];
        setCurrentPath((currentPath) => {
          // first we build up the "new path" according to whatever "from" mode
          // we're in
          let newPath: AnyStep[];
          if (from === "root" || from === "here") {
            newPath = toArray;
          } else {
            const fromIndex = currentPath.findLastIndex(
              (step) => step.direction === from,
            );
            if (fromIndex === -1) {
              throw new Error(`Cannot navigate from ${from.description}`);
            }
            return [...currentPath.slice(0, fromIndex + 1), ...toArray];
          }
          if (to === "up") {
            if (newPath.length === 0) {
              throw new Error("Cannot navigate up from root (says Router)");
            }
            newPath = currentPath.slice(0, currentPath.length - 1);
          }
          // now we map the newpath, and if any step elements match the old
          // path in all but id, we replace them with the old one.
          // we do this because steps are recreated on every render, so they can
          // have different ids, and we want navigation to remain stable.
          newPath = newPath.map((step, i) => {
            if (
              currentPath[i] &&
              step.direction === currentPath[i].direction &&
              (step.params === currentPath[i].params ||
                deepEquals(step.params, currentPath[i].params))
            ) {
              return currentPath[i];
            } else {
              return step;
            }
          });
          return newPath;
        });
      },
      currentStep: currentPath[0],
      parentSteps: [],
      childSteps: currentPath.slice(1),
    }),
    [currentPath],
  );
  return (
    <NavigationContext.Provider value={navigationContextValue}>
      {children}
    </NavigationContext.Provider>
  );
};
