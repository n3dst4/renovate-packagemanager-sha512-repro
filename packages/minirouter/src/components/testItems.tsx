import React from "react";

import { createDirection } from "../createDirection";
import { Outlet } from "../outlets/Outlet";
import { useParams } from "../useParams";
import { Link } from "./Link";
import { Route } from "./Route";
import { Router } from "./Router";

export const direction1 = createDirection("foo");
export const about = createDirection("about");
export const values = createDirection("values");
export const team = createDirection("team");
export const member = createDirection<string>("member");
export const contact = createDirection("contact");
export const email = createDirection("email");
export const phone = createDirection("phone");
export const social = createDirection("social");
export const categories = createDirection("categories");
export const category = createDirection<number>("category");
export const product = createDirection<{
  id: number;
  variant: string;
}>("product");

const Values = () => (
  <div data-testid="values-page">
    <h2>Values</h2>
    Our values are important to our{" "}
    <Link from={about} to={team()}>
      team
    </Link>
    and you can
    <Link to={contact()}>contact us</Link>
    about them.
  </div>
);

const Product = () => {
  const { id, variant } = useParams(product);
  return (
    <div data-testid={`product-${id}-${variant}-page`}>
      <h2>
        This is product {id} in {variant}
      </h2>
      <Link to="up">back to category</Link>
    </div>
  );
};

Product.displayName = "Product";

const Category = () => {
  const categoryId = useParams(category);
  return (
    <div data-testid={`category-${categoryId}-page`}>
      <h2>Category {categoryId}</h2>
      <Link to="up">back to categories</Link>
      <Link
        from={category}
        to={product({
          id: 1,
          variant: "beige",
        })}
      >
        Product 1 in beige
      </Link>
      <Link
        from={category}
        to={product({
          id: 1,
          variant: "blue",
        })}
      >
        Product 1 in blue
      </Link>
      <Link
        from={category}
        to={product({
          id: 2,
          variant: "red",
        })}
      >
        Product 2 in red
      </Link>
      <Link
        from={category}
        to={product({
          id: 3,
          variant: "yellow",
        })}
      >
        Product 3 in yellow
      </Link>
      <Link
        from={category}
        to={product({
          id: 3,
          variant: "green",
        })}
      >
        Product 3 in green
      </Link>
      <Route direction={product}>
        <Product />
      </Route>
    </div>
  );
};

const Categories = () => (
  <div data-testid="categories-page">
    <h2>Categories</h2>
    <ul>
      <li>
        <Link from={categories} to={category(1)}>
          Category 1
        </Link>
      </li>
      <li>
        <Link from={categories} to={category(2)}>
          Category 2
        </Link>
      </li>
      <li>
        <Link from={categories} to={category(3)}>
          Category 3
        </Link>
      </li>
    </ul>
    <Route direction={category}>
      <Category />
    </Route>
  </div>
);

const Member = () => {
  const name = useParams(member);
  return (
    <div>
      <h2>{name} is a member of our team</h2>
      <Link to="up">back to team</Link>
    </div>
  );
};

const Contact = () => (
  <div data-testid="contact-page">
    <h2>Contact us</h2>
    <Link to={email()}>Email us</Link>
    <Link to={phone()}>Phone us</Link>
    <Link to={social()}>Social media</Link>
    <Outlet>
      <Route direction={email}>
        <div>foo@example.com</div>
      </Route>
      <Route direction={phone}>
        <div>01234 567890</div>
      </Route>
      <Route direction={social}>
        <div>@wobble on Wibble</div>
      </Route>
    </Outlet>
  </div>
);

export const App = () => (
  <Router>
    <h1>Terrific company website</h1>
    {/* top level nav */}
    <Link to={about()}>About</Link>
    <Link from="root" to={contact()}>
      Contact
    </Link>
    <Link to={[about(), values()]}>Values</Link>
    <Link to={[categories()]}>Product Categories</Link>5
    <Route direction={about}>
      <div data-testid="about-page" className="about">
        <h2>About</h2>
        <p>Learn more about our</p>
        <ul>
          <li>
            <Link to={values()}>Values</Link>
          </li>
          <li>
            <Link to={team()}>Team</Link>
          </li>
        </ul>

        <Route direction={values}>
          <Values />
        </Route>

        <Route direction={team}>
          <div data-testid="team-page">
            <h2>Team</h2>
            Our team is great
            <Link to={member("Alice")}>Alice</Link>
            <Link to={member("Bob")}>Bob</Link>
            <Link to={member("Carla")}>Carla</Link>
            <Link from="root" to={[]}>
              Home
            </Link>
          </div>
          <Route direction={member}>
            <Member />
          </Route>
        </Route>
      </div>
    </Route>
    <Route direction={contact}>
      <Contact />
    </Route>
    <Route direction={categories}>
      <Categories />
    </Route>
  </Router>
);
