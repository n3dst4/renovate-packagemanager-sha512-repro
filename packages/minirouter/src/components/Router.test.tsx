import "@testing-library/jest-dom/vitest";

import { getByTestId, getByText, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { describe, expect, test } from "vitest";

import { Link } from "./Link";
import { Route } from "./Route";
import { Router } from "./Router";
import { App, direction1 } from "./testItems";

const user = userEvent.setup();

describe("Empty", () => {
  test("should render nothing", () => {
    const { container } = render(<Router />);
    expect(container.innerHTML).toBe("");
    expect(true).toBe(true);
  });
});

describe("root content", () => {
  test("should render content", () => {
    const { container } = render(
      <Router>
        <p>root</p>
      </Router>,
    );
    expect(container.innerHTML).toContain("root");
  });
});

describe("single route", () => {
  test("should not show content immediately", () => {
    const { container } = render(
      <Router>
        <Route direction={direction1}>
          <p>route</p>
        </Route>
      </Router>,
    );
    expect(container.innerHTML).not.toContain("route");
  });

  test("should show content when a link is clicked", async () => {
    const { container } = render(
      <Router>
        <Link data-testid="link" from="root" to={direction1()}></Link>
        <Route direction={direction1}>
          <p>route</p>
        </Route>
      </Router>,
    );
    expect(container.innerHTML).not.toContain("route");
    const link = container.querySelector("a[data-testid=link]");

    expect(link).not.toBeNull();

    await user.click(link!);
    expect(container.innerHTML).toContain("route");
  });
});

test("navigation story", async () => {
  // 1 render the router
  const { container } = render(<App />);
  expect(container).toMatchSnapshot("001 initial state");
  expect(getByText(container, "Terrific company website")).toBeInTheDocument();
  expect(getByText(container, "About", { selector: "a" })).toBeInTheDocument();
  expect(
    getByText(container, "Contact", { selector: "a" }),
  ).toBeInTheDocument();
  expect(getByText(container, "Values", { selector: "a" })).toBeInTheDocument();

  // 2 visit the about page
  await user.click(getByText(container, "About", { selector: "a" }));
  expect(container).toMatchSnapshot("002 about page");
  const aboutPage = getByTestId(container, "about-page");
  const innerValuesLink = getByText(aboutPage, "Values", { selector: "a" });
  getByText(container, "Team", { selector: "a" });

  // 3 visit the values page
  await user.click(innerValuesLink);
  expect(container).toMatchSnapshot("003 values page");
  const valuesPage = getByTestId(container, "values-page");
  expect(container).toHaveTextContent("Our values are important to our team");
  // expect(getByText("Team", { selector: "a" })).toBeInTheDocument();
  const valuesLinkToTeam = getByText(valuesPage, "team", { selector: "a" });

  // 4 go sideways to team
  await user.click(valuesLinkToTeam);
  expect(container).toMatchSnapshot("004 team page");
  expect(container).toHaveTextContent("Our team is great");
  expect(container).not.toHaveTextContent(
    "Our values are important to our team",
  );
  const teamPage = getByTestId(container, "team-page");

  // 5 go to Members' pages
  await user.click(getByText(teamPage, "Alice", { selector: "a" }));
  expect(container).toMatchSnapshot("005 alice page");
  expect(container).toHaveTextContent("Alice is a member of our team");
  await user.click(getByText(teamPage, "Bob", { selector: "a" }));
  expect(container).toHaveTextContent("Bob is a member of our team");
  expect(container).not.toHaveTextContent("Alice is a member of our team");
  await user.click(getByText(teamPage, "Carla", { selector: "a" }));
  expect(container).toHaveTextContent("Carla is a member of our team");
  expect(container).not.toHaveTextContent("Alice is a member of our team");
  expect(container).not.toHaveTextContent("Bob is a member of our team");

  // 6 go home
  await user.click(getByText(teamPage, "Home", { selector: "a" }));
  expect(container).toMatchSnapshot("006 home page");
  expect(container).not.toHaveTextContent("Our team is great");
  expect(container).not.toHaveTextContent("Learn more about our");
  expect(container).not.toHaveTextContent("Alice is a member of our team");
  expect(container).not.toHaveTextContent("Bob is a member of our team");
  expect(container).not.toHaveTextContent("Carla is a member of our team");

  // 7 go to contact
  await user.click(getByText(container, "Contact", { selector: "a" }));
  expect(container).toMatchSnapshot("007 contact page");
  expect(container).toHaveTextContent("Contact us");
  expect(container).not.toHaveTextContent("foo@example.com");
  expect(container).not.toHaveTextContent("01234 567890");
  expect(container).not.toHaveTextContent("@wobble on Wibble");
  const contactPage = getByTestId(container, "contact-page");
  await user.click(getByText(contactPage, "Email us", { selector: "a" }));
  expect(container).toHaveTextContent("foo@example.com");
  expect(container).not.toHaveTextContent("01234 567890");
  expect(container).not.toHaveTextContent("@wobble on Wibble");
  await user.click(getByText(contactPage, "Phone us", { selector: "a" }));
  expect(container).not.toHaveTextContent("foo@example.com");
  expect(container).toHaveTextContent("01234 567890");
  expect(container).not.toHaveTextContent("@wobble on Wibble");
  await user.click(getByText(contactPage, "Social media", { selector: "a" }));
  expect(container).toHaveTextContent("@wobble on Wibble");
  expect(container).not.toHaveTextContent("foo@example.com");
  expect(container).not.toHaveTextContent("01234 567890");

  // 8 go to categories
  await user.click(
    getByText(container, "Product Categories", { selector: "a" }),
  );
  expect(container).not.toHaveTextContent("Contact us");
  const categoriesPage = getByTestId(container, "categories-page");
  expect(categoriesPage).toHaveTextContent("Categories");
  await user.click(getByText(categoriesPage, "Category 1", { selector: "a" }));
  const category1Page = getByTestId(container, "category-1-page");
  expect(category1Page).toHaveTextContent("Category 1");
  await user.click(
    getByText(category1Page, "Product 1 in beige", { selector: "a" }),
  );
  expect(container).toHaveTextContent("This is product 1 in beige");
  expect(container).not.toHaveTextContent("This is product 1 in blue");
  await user.click(
    getByText(category1Page, "Product 1 in blue", { selector: "a" }),
  );
  expect(container).toHaveTextContent("This is product 1 in blue");
  expect(container).not.toHaveTextContent("This is product 1 in beige");
});
