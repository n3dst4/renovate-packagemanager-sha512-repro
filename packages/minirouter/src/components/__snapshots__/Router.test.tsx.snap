// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`navigation story > 001 initial state 1`] = `
<div>
  <h1>
    Terrific company website
  </h1>
  <a
    tabindex="0"
  >
    About
  </a>
  <a
    tabindex="0"
  >
    Contact
  </a>
  <a
    tabindex="0"
  >
    Values
  </a>
  <a
    tabindex="0"
  >
    Product Categories
  </a>
  5
</div>
`;

exports[`navigation story > 002 about page 1`] = `
<div>
  <h1>
    Terrific company website
  </h1>
  <a
    tabindex="0"
  >
    About
  </a>
  <a
    tabindex="0"
  >
    Contact
  </a>
  <a
    tabindex="0"
  >
    Values
  </a>
  <a
    tabindex="0"
  >
    Product Categories
  </a>
  5
  <div
    class="about"
    data-testid="about-page"
  >
    <h2>
      About
    </h2>
    <p>
      Learn more about our
    </p>
    <ul>
      <li>
        <a
          tabindex="0"
        >
          Values
        </a>
      </li>
      <li>
        <a
          tabindex="0"
        >
          Team
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`navigation story > 003 values page 1`] = `
<div>
  <h1>
    Terrific company website
  </h1>
  <a
    tabindex="0"
  >
    About
  </a>
  <a
    tabindex="0"
  >
    Contact
  </a>
  <a
    tabindex="0"
  >
    Values
  </a>
  <a
    tabindex="0"
  >
    Product Categories
  </a>
  5
  <div
    class="about"
    data-testid="about-page"
  >
    <h2>
      About
    </h2>
    <p>
      Learn more about our
    </p>
    <ul>
      <li>
        <a
          tabindex="0"
        >
          Values
        </a>
      </li>
      <li>
        <a
          tabindex="0"
        >
          Team
        </a>
      </li>
    </ul>
    <div
      data-testid="values-page"
    >
      <h2>
        Values
      </h2>
      Our values are important to our
       
      <a
        tabindex="0"
      >
        team
      </a>
      and you can
      <a
        tabindex="0"
      >
        contact us
      </a>
      about them.
    </div>
  </div>
</div>
`;

exports[`navigation story > 004 team page 1`] = `
<div>
  <h1>
    Terrific company website
  </h1>
  <a
    tabindex="0"
  >
    About
  </a>
  <a
    tabindex="0"
  >
    Contact
  </a>
  <a
    tabindex="0"
  >
    Values
  </a>
  <a
    tabindex="0"
  >
    Product Categories
  </a>
  5
  <div
    class="about"
    data-testid="about-page"
  >
    <h2>
      About
    </h2>
    <p>
      Learn more about our
    </p>
    <ul>
      <li>
        <a
          tabindex="0"
        >
          Values
        </a>
      </li>
      <li>
        <a
          tabindex="0"
        >
          Team
        </a>
      </li>
    </ul>
    <div
      data-testid="team-page"
    >
      <h2>
        Team
      </h2>
      Our team is great
      <a
        tabindex="0"
      >
        Alice
      </a>
      <a
        tabindex="0"
      >
        Bob
      </a>
      <a
        tabindex="0"
      >
        Carla
      </a>
      <a
        tabindex="0"
      >
        Home
      </a>
    </div>
  </div>
</div>
`;

exports[`navigation story > 005 alice page 1`] = `
<div>
  <h1>
    Terrific company website
  </h1>
  <a
    tabindex="0"
  >
    About
  </a>
  <a
    tabindex="0"
  >
    Contact
  </a>
  <a
    tabindex="0"
  >
    Values
  </a>
  <a
    tabindex="0"
  >
    Product Categories
  </a>
  5
  <div
    class="about"
    data-testid="about-page"
  >
    <h2>
      About
    </h2>
    <p>
      Learn more about our
    </p>
    <ul>
      <li>
        <a
          tabindex="0"
        >
          Values
        </a>
      </li>
      <li>
        <a
          tabindex="0"
        >
          Team
        </a>
      </li>
    </ul>
    <div
      data-testid="team-page"
    >
      <h2>
        Team
      </h2>
      Our team is great
      <a
        tabindex="0"
      >
        Alice
      </a>
      <a
        tabindex="0"
      >
        Bob
      </a>
      <a
        tabindex="0"
      >
        Carla
      </a>
      <a
        tabindex="0"
      >
        Home
      </a>
    </div>
    <div>
      <h2>
        Alice
         is a member of our team
      </h2>
      <a
        tabindex="0"
      >
        back to team
      </a>
    </div>
  </div>
</div>
`;

exports[`navigation story > 006 home page 1`] = `
<div>
  <h1>
    Terrific company website
  </h1>
  <a
    tabindex="0"
  >
    About
  </a>
  <a
    tabindex="0"
  >
    Contact
  </a>
  <a
    tabindex="0"
  >
    Values
  </a>
  <a
    tabindex="0"
  >
    Product Categories
  </a>
  5
</div>
`;

exports[`navigation story > 007 contact page 1`] = `
<div>
  <h1>
    Terrific company website
  </h1>
  <a
    tabindex="0"
  >
    About
  </a>
  <a
    tabindex="0"
  >
    Contact
  </a>
  <a
    tabindex="0"
  >
    Values
  </a>
  <a
    tabindex="0"
  >
    Product Categories
  </a>
  5
  <div
    data-testid="contact-page"
  >
    <h2>
      Contact us
    </h2>
    <a
      tabindex="0"
    >
      Email us
    </a>
    <a
      tabindex="0"
    >
      Phone us
    </a>
    <a
      tabindex="0"
    >
      Social media
    </a>
  </div>
</div>
`;
